README.md

Apns4erl v2 [![Build Status](https://travis-ci.org/inaka/apns4erl.svg?branch=master)](https://travis-ci.org/inaka/apns4erl)[![codecov](https://codecov.io/gh/inaka/apns4erl/branch/master/graph/badge.svg)](https://codecov.io/gh/inaka/apns4erl)
========

<img src="https://media.giphy.com/media/uZQP0PR0BmkGA/giphy.gif" align="right" style="float:right" height="400" />

This lib is intended to allow you to write an APNs provider for [Apple Push Notificaion services (APNs)](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html) over HTTP2 in Erlang.

Copyright (c) 2017 Erlang Solutions Ltd. <support@inaka.net>, released under the Apache 2 license

You can find the v1 [here](https://github.com/inaka/apns4erl/releases/tag/1.0.6-final)?

__Note:__ `Apns4erl v2` is still under development. Currently it supports push notifications with certificate and authentication token.

Contact Us
==========
If you find any **bugs** or have a **problem** while using Apns4erl, please [open an issue](https://github.com/inaka/apns4erl/issues/new) in this repo (or a pull request :)).

## Requirements
- You must have installed an updated Openssl version or, at least, be sure it supports TLS 1.2+. New APNs server only supports connections over TLS 1.2+.
- Erlang R19+

## Important Links

- [Pool of connections Example](examples/apns_pool/README.md)

## How to use it?

First we have to fill our `config` data. There are two ways for do this, one is filling a `config` file. This is an example you can find at `test/test.config`:

```
{
  apns,
  [ {apple_host,       "api.development.push.apple.com"}
  , {apple_port,       443}
  , {certfile,         "priv/apns-dev-cert.pem"}
  , {keyfile,          "priv/apns-dev-key-noenc.pem"}
  , {token_keyfile,    "priv/APNsAuthKey_KEYID12345.p8"}
  , {timeout,          10000}

  %% APNs Headers

  , {apns_id,          undefined}
  , {apns_expiration,  0}
  , {apns_priority,    10}
  , {apns_topic,       "com.example.myapp"}
  , {apns_collapse_id, undefined}
  , {apns_push_type,   "alert"}

  %% Feedback
  , {feedback_host,    "feedback.push.apple.com"}
  , {feedback_port,    2195}
  ]
  ]
}
```

The other way is send all that info as a parameter to `apns:connect/1` function encapsulated in a `apns_connection:connection()` structure:

```erlang
#{ name       := name()
 , apple_host := host()
 , apple_port := inet:port_number()
 , certfile   => path()
 , keyfile    => path()
 , timeout    => integer()
 , type       := type()
 }.
```

APNs allows two connection types, one is using `Provider Certificates`. The first certificate option is to supply cert paths in `certfile` and `keyfile`. Alternatively, you can supply a cert binary in `certdata` and a `keydata()`-type tuple (see: https://github.com/inaka/apns4erl/blob/master/src/apns_connection.erl#L64) in `keydata`. Certs are the `Provider Certificates` and the keys are the `Private Key` both provided by Apple. We need them in `.pem` format, here is an example of how to convert them, check the [certificates](https://blog.serverdensity.com/how-to-build-an-apple-push-notification-provider-server-tutorial/) section.

The other way to connect against APNs is using `Provider Authentication Tokens`, for this choice you must fill the field `token_keyfile`. This is a path to the Authentication Key provided by Apple. This is in `.p8` format and it doesn't need conversion.

This `key` will be needed in order to generate a token which will be used every time we try to push a notification. In connection's time it is not needed.

## Run

`apns4erl` can be included as a dependency and started from `yourapp.app.src`. You also can run it on the shell for testing.

```
> rebar3 compile
> erl -pa _build/default/lib/*/ebin -config test/test.config
```
Don't forget your config file if you want to use `apns:connect/2`.
```erlang
1> apns:start().
ok
```

## Create connections

After running `apns4erl` app we can start creating connections. As we mentioned before there are two types of connections. Both are created using the functions `apns:connect/1` and `apns:connect/2`.

- `apns:connect/1`: This function accepts as a parameter an `apns_connection:connection()` structure.
  ```erlang
  #{ name       := name()
   , apple_host := host()
   , apple_port := inet:port_number()
   , certdata   => binary()
   , certfile   => path()
   , keydata    => keydata()
   , keyfile    => path()
   , timeout    => integer()
   , type       := type()
   }.
  ```
  where the `type` field indicates if is `certdata`, `cert`, or `token`.

- `apns:connect/2`: The first argument is the type and the second one is the connection's name. In order to use it successfully we have to fill the `config` file before, as explained in `how to use it?` section.

Example:

```erlang
1> apns:connect(cert, my_first_connection).
{ok,<0.87.0>}
2> apns:connect(#{name => another_cert, apple_host => "api.push.apple.com", apple_port => 443,
certfile => "priv/cert.pem", keyfile => "priv/key.pem", type => cert}).
3> apns:connect(token, my_second_connection).
{ok,<0.95.0>}
```
Note `cert` and `token` define the type we want.

`apns:connect/2` returns the connection `pid`.

## Create Connections without name

In some scenarios we don't want to assign names to the connections instead we want working just with the `pid` (working with a pool of connections for example). If that is the case we use the same `apns:connect/1` and `apns:connect/2` functions but instead of a connection name we put `undefined`:

```erlang
1> apns:connect(cert, undefined).
{ok,<0.127.0>}
2> apns:connect(#{name => undefined, apple_host => "api.push.apple.com", apple_port => 443,
certfile => "priv/cert2.pem", keyfile => "priv/key2-noenc.pem", type => cert}).
{ok,<0.130.0>}
3> apns:connect(token, my_second_connection).
{ok,<0.132.0>}
```

## Push Notifications over `Provider Certificate` connections

In order to send Notifications over `Provider Certificate` connection we will use `apns:push_notification/3,4`.

We will need the connection, a notification, the device ID and some http2 headers. The connection is the `atom` we used when we executed `apns:connect/2` for setting a name or its `pid`, the device ID is provided by Apple, the notification is a `map` with the data we want to send, that map will be encoded to json later and the http2 headers can be explicitly sent as a parameter using `apns:push_notification/4` or can be defined at the `config` file, in that case we would use `apns:push_notification/3`.

This is the `headers` format:

```erlang
-type headers()   :: #{ apns_id          => binary()
                      , apns_expiration  => binary()
                      , apns_priority    => binary()
                      , apns_topic       => binary()
                      , apns_collapse_id => binary()
                      , apns_push_type   => binary()
                      , apns_auth_token  => binary()
                      }.
```

All of them are defined by Apple  [here](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CommunicatingwithAPNs.html)

Lets send a Notification.

```erlang
1> {ok, Pid} = apns:connect(cert, my_first_connection).
{ok,<0.85.0>}
2> DeviceId = <<"a0dc63fb059cb9c13b03e5c974af3dd33d67fed4147da8c5ada0626439e18935">>.
<<"a0dc63fb059cb9c13b03e5c974af3dd33d67fed4147da8c5ada0626439e18935">>
3> Notification = #{aps => #{alert => <<"you have a message">>}}.
#{aps => #{alert => <<"you have a message">>}}
4> apns:push_notification(my_first_connection, DeviceId, Notification).
{200,
 [{<<"apns-id">>,<<"EFDE0D9D-F60C-30F4-3FF1-86F3B90BE434">>}],
 no_body}
5> apns:push_notification(Pid, DeviceId, Notification).
{200,
 [{<<"apns-id">>,<<"EFDE0D9D-F60C-30F4-3FF1-86F3B90BE654">>}],
 no_body}
```

The result is the response itself, its format is:

```erlang
-type response()  :: { integer()          % HTTP2 Code
                     , [term()]           % Response Headers
                     , [term()] | no_body % Response Body
                     } | timeout.
```

And that's all.

## Push Notifications over `Provider Authentication Tokens` connections

This is the other way APNs allows us to send notifications. In this case we don't need a certificate but we will need a `p8` file with the private key we will use to sign the token. Lets assume we've got the file  `APNsAuthKey_KEYID12345.p8` from Apple. We then have to fill the `config` file key `token_keyfile` with the path to that file.

We will need a `kid` value, this is the key identifier. In our case is the last 10 chars of the file name (`KEYID123456`). We will need also the `iss` value, this is the Team Id, that can be checked on your Apple's Developer account, in our case it will be `THEATEAM`. And that's it.

You can find more info [here](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CommunicatingwithAPNs.html)

In order to push a notification we will use `apns:push_notification_token/4,5`. We will need the same attributes we used sending a notification over `Provider Certificate` connections plus a signed `token`. This token has a 1 hour life, so that means we can generate one token and use it many times till it expires. Lets try.

Create the token:


CHANGELOG.md
# Changelog

## [2.4.0](https://github.com/inaka/apns4erl/tree/2.4.0) (2020-08-23)

[Full Changelog](https://github.com/inaka/apns4erl/compare/e987f52410f3572de61b8610b677338881582656...2.4.0)

**Closed issues:**

- version bump to 2.3.1 [\#227](https://github.com/inaka/apns4erl/issues/227)
- Rebar compile issues when included in other projects [\#225](https://github.com/inaka/apns4erl/issues/225)
- how to send mutiple token to apns in one time? [\#223](https://github.com/inaka/apns4erl/issues/223)
- Connections using Provider Auth Token [\#222](https://github.com/inaka/apns4erl/issues/222)
- adding coverage in Travis [\#218](https://github.com/inaka/apns4erl/issues/218)
- Version bump to 2.3.0 [\#217](https://github.com/inaka/apns4erl/issues/217)
- why push\_notification\_token return "error" [\#214](https://github.com/inaka/apns4erl/issues/214)
- Pseudo-header field ':method' found after regular header [\#211](https://github.com/inaka/apns4erl/issues/211)
- apns:close\_connection don't closed h2\_connect [\#209](https://github.com/inaka/apns4erl/issues/209)
- Memory leaks may exist [\#206](https://github.com/inaka/apns4erl/issues/206)
- Version Bump to 2.2.0 [\#196](https://github.com/inaka/apns4erl/issues/196)
- Add Travis [\#193](https://github.com/inaka/apns4erl/issues/193)
- Write an example with a pool of connections [\#191](https://github.com/inaka/apns4erl/issues/191)
- restrict calls to push\_notification and push\_notification\_token [\#188](https://github.com/inaka/apns4erl/issues/188)
- Version Bump to 2.1.1 [\#186](https://github.com/inaka/apns4erl/issues/186)
- chatterbox OPT20 gen\_fsm is deprecated and will be removed in a future release [\#184](https://github.com/inaka/apns4erl/issues/184)
- Create connections without name [\#183](https://github.com/inaka/apns4erl/issues/183)
- gen\_server push\_notification return {Timeout, StreamId} [\#182](https://github.com/inaka/apns4erl/issues/182)
- Base64 newline characters [\#180](https://github.com/inaka/apns4erl/issues/180)
- Version Bump to 2.1.0 [\#177](https://github.com/inaka/apns4erl/issues/177)
- research in HTTP2 clients [\#175](https://github.com/inaka/apns4erl/issues/175)
- V2 Need async\_push\_notification [\#172](https://github.com/inaka/apns4erl/issues/172)
- apns:push\_notification timeout [\#170](https://github.com/inaka/apns4erl/issues/170)
- Erlang Crashing [\#168](https://github.com/inaka/apns4erl/issues/168)
- implement apns:get\_feedback/1 [\#167](https://github.com/inaka/apns4erl/issues/167)
- where do I register feedback handlers in v2? [\#166](https://github.com/inaka/apns4erl/issues/166)
- Syntax error rebar3 compile [\#165](https://github.com/inaka/apns4erl/issues/165)
- Erlang/OTP 18? [\#164](https://github.com/inaka/apns4erl/issues/164)
- How to find out the invalid token? [\#163](https://github.com/inaka/apns4erl/issues/163)
- Improve `apns\_utils:sing/1` in order to use `os:cmd/1` instead of `ktn\_os:command/1` [\#159](https://github.com/inaka/apns4erl/issues/159)
- add apns:connection/1 documentation to README [\#156](https://github.com/inaka/apns4erl/issues/156)
- update katana dependency [\#153](https://github.com/inaka/apns4erl/issues/153)
- rebar3 compile error [\#152](https://github.com/inaka/apns4erl/issues/152)
- upload to hex [\#151](https://github.com/inaka/apns4erl/issues/151)
- test coverage [\#148](https://github.com/inaka/apns4erl/issues/148)
- Update README and rebar in order to support only R19+ [\#146](https://github.com/inaka/apns4erl/issues/146)
- version bump to 2.0.0 [\#145](https://github.com/inaka/apns4erl/issues/145)
- Handling responses improvement [\#138](https://github.com/inaka/apns4erl/issues/138)
- Update Documentation [\#136](https://github.com/inaka/apns4erl/issues/136)
- Push Notifications with Provider's Certificate [\#135](https://github.com/inaka/apns4erl/issues/135)
- close connection gracefully [\#133](https://github.com/inaka/apns4erl/issues/133)
- Get Feedback from APNs [\#132](https://github.com/inaka/apns4erl/issues/132)
- Exponential Backoff when restarting gun connection [\#131](https://github.com/inaka/apns4erl/issues/131)
- apns\_connection should monitor the gun\_connection [\#129](https://github.com/inaka/apns4erl/issues/129)
- Create a Queue for tasks [\#127](https://github.com/inaka/apns4erl/issues/127)
- Create Connections using Provider Auth Token [\#126](https://github.com/inaka/apns4erl/issues/126)
- Create Connections using Provider Certificate [\#125](https://github.com/inaka/apns4erl/issues/125)

erldocs.js
var ErlDocs = (function(index) {

    /* Search box height? */
    var RESULT_OFFSET_Y = 26;

    /* Human readable keyCode index */
    var KEY = {'BACKSPACE': 8, 'TAB': 9, 'NUM_PAD_CLEAR': 12, 'ENTER': 13, 'SHIFT': 16, 'CTRL': 17, 'ALT': 18, 'PAUSE': 19, 'CAPS_LOCK': 20, 'ESCAPE': 27, 'SPACEBAR': 32, 'PAGE_UP': 33, 'PAGE_DOWN': 34, 'END': 35, 'HOME': 36, 'ARROW_LEFT': 37, 'ARROW_UP': 38, 'ARROW_RIGHT': 39, 'ARROW_DOWN': 40, 'PRINT_SCREEN': 44, 'INSERT': 45, 'DELETE': 46, 'SEMICOLON': 59, 'WINDOWS_LEFT': 91, 'WINDOWS_RIGHT': 92, 'SELECT': 93, 'NUM_PAD_ASTERISK': 106, 'NUM_PAD_PLUS_SIGN': 107, 'NUM_PAD_HYPHEN-MINUS': 109, 'NUM_PAD_FULL_STOP': 110, 'NUM_PAD_SOLIDUS': 111, 'NUM_LOCK': 144, 'SCROLL_LOCK': 145, 'SEMICOLON': 186, 'EQUALS_SIGN': 187, 'COMMA': 188, 'HYPHEN-MINUS': 189, 'FULL_STOP': 190, 'SOLIDUS': 191, 'GRAVE_ACCENT': 192, 'LEFT_SQUARE_BRACKET': 219, 'REVERSE_SOLIDUS': 220, 'RIGHT_SQUARE_BRACKET': 221, 'APOSTROPHE': 222};

    (function () {
        /* 0 - 9 */
	for (var i = 48; i <= 57; i++) {
            KEY['' + (i - 48)] = i;
	}
	/* A - Z */
	for (i = 65; i <= 90; i++) {
            KEY['' + String.fromCharCode(i)] = i;
	}
	/* NUM_PAD_0 - NUM_PAD_9 */
	for (i = 96; i <= 105; i++) {
            KEY['NUM_PAD_' + (i - 96)] = i;
	}
	/* F1 - F12 */
	for (i = 112; i <= 123; i++) {
            KEY['F' + (i - 112 + 1)] = i;
	}
    })();

    var $search       = $("#search");
    var $results      = $("#results");
    var searchActive  = false;
    var selected      = null;
    var resultsCount  = 0;
    // var showingFuns   = true;


    // var getDetails = function() {
    //     return showingFuns ? {"text":"Hide Functions", "cBottom": "25%", fTop: "75%"}
    //                        : {"text":"View Functions", "cBottom": "0px", fTop: "100%"};
    // };

    // var setDetails = function(details, fun) {
    //     $("#viewfuns").text(details.text);
    //     $("#content").css({"bottom":details.cBottom});
    //     $("#funwrapper").animate({"top":details.fTop}, "normal", fun);
    // };

    function scrollIntoView($parent, $child, force) {
        var childTop     = $child.position().top - RESULT_OFFSET_Y;
        var scrollTop    = $parent[0].scrollTop;
        var childHeight  = $child.height();
        var parentHeight = $parent.height();
        var newTop       = null;

	if (force) {
            $parent[0].scrollTop = scrollTop + childTop;
            return;
        }

        if (childTop < 0) {
            newTop = scrollTop + childTop;
        } else if (childTop + childHeight > parentHeight) {
            newTop = scrollTop + ((childTop + childHeight) - parentHeight);
        }

        if (newTop !== null) {
            if (Math.abs(newTop - scrollTop) > 200) {
                $parent.animate({"scrollTop": newTop}, 'fast');
            } else {
                $parent[0].scrollTop = newTop;
            }
        }
    }

    function setSelected(x, force) {
        var sel, children = $results.children("li");

        if (x >= 0 && x < resultsCount) {
            if (selected !== null) {
                children.eq(selected).removeClass("selected");
            }
            selected = x;
            sel = children.eq(x).addClass("selected");
            if (sel.length > 0) {
                scrollIntoView($results, sel, force);
            }
        }
    }

    function keypress(e) {
        var tmp, blockedKeys = [0, KEY.TAB, KEY.CTRL, KEY.ALT,
                                KEY.WINDOWS_LEFT, KEY.GRAVE_ACCENT];

        if (!($.inArray(e.keyCode, blockedKeys) === -1)) {
            return;
        }

             if (e.keyCode === KEY.ARROW_DOWN) setSelected(selected + 1);
        else if (e.keyCode === KEY.ARROW_UP)   setSelected(selected - 1);
        else if (e.keyCode === KEY.PAGE_DOWN)  setSelected(selected + 7);
        else if (e.keyCode === KEY.PAGE_UP)    setSelected(selected - 7);
        else if (e.keyCode === KEY.ENTER) {
            tmp = $results.children(".selected");
            if (tmp.length > 0) {
                document.location.href = tmp.find("a").attr("href");
            }
        } else {
            filter($search.val());
        }
    }

    function windowResize() {
        $results.height($(window).height() - RESULT_OFFSET_Y);
    }

    function showModules() {
        var i, item, len = index.length, results = [];

        for (i = 0; i < len; i += 1) {
            item = index[i];
            if (item[0] === "mod" || item[0] === "app") {
                results.push(item);
            }
        }
        return results;
    }

    function searchApps(str) {
        var i, count, item;
        var len     = index.length;
        var results = [];
        var terms   = str.split(" ");

        for (i = 0, count = 0; i < len; i += 1) {
            item = index[i];
            if (match(item[2], terms)) {
                results.push(item);
                if (count++ > 100) {
                    break;
                }
            }
        }
        return results;
    }

    function formatResults(results, str) {
        var i, item, hash, url, path, html = "";
        var len       = results.length;
        var searchStr = isSearchStr(str) ? "&search="+str : "";

        for (i = 0; i < len; i += 1) {

            item = results[i];

            if (item[0] === "app") {
                hash = "#" + item[1];
                path = "index";
            } else {
                hash = (item[0] === "fun") ? "#" + item[2].split(":")[1] : "";
                path = item[1] + "/" + item[2].split(":")[0];
            }
            url = CURRENT_ROOT + path + ".html?i=" + i + searchStr + hash;

            html += '<li class="'+ item[0] +'"><a href="'+ url +'">' +
                '<span class="name">'+ item[2]       +"</span><br/>" +
                '<span class="sub">'+  item[3]       +'</span>'      +
                '</a></li>';
        }

        return html;
    }

    function isSearchStr(str) {
        return typeof str !== "undefined" && str !== "";
    }

    function filter(str) {
        var results = isSearchStr(str) ? searchApps(str) : showModules();
        var html    = formatResults(results, str);

        $results[0].innerHTML = html;
        setSelected(0);
        resultsCount = results.length;
    }

    function match(str, terms) {
        for (var i = 0, len = terms.length; i < len; i += 1) {
            if (str.match(new RegExp(terms[i], "i")) === null) {
                return false;
            }
        }
        return true;
    }

    function parseQuery(url) {
        var arr, query, i, len, tmp;
        var qs = url.split("?")[1];

        if (typeof qs !== "undefined") {
            arr   = qs.split("&");
            query = {};
            for (i = 0, len = arr.length; i < len; i += 1) {
                tmp = arr[i].split("=");
                query[tmp[0]] = tmp[1];
            }
            return query;
        }
        return false;
    }

    function strToBool(val) {
        return val === "true";
    }

    function updateTitle() {
        var results = showModules();
        var len = results.length;
        var $t = $("body h1");
        var mod = $t.text();

        for (var i = 0; i < len; i++) {
            var idx = results[i];
            if (idx[0] !== "mod" || idx[2] !== mod)
                continue;
            var app = idx[1];
            for (var j = i - 1; j >= 0; j--) {
                idx = results[j];
                if (idx[0] !== "app" || idx[1] !== app)
                    continue;
                $t.after(' (<a href="../index.html?i=' + j + '#' + app + '">' + app + '</a>)');
                return;
            }
        }
    }

    function init() {
        var val, qs = parseQuery(document.location.search);

        if (qs && qs.search && $search.val() === "") {
            val = decodeURIComponent(qs.search.replace(/\+/g,  " "));
            $search.val(val);
            filter(val);

            // If a search parameter is specified and hash doesnt already
            // exist, jump to first match
            if (document.location.hash === "") {
                document.location.hash =
                    $("#results li:first-child a").attr("href").split("#")[1];
            }

        } else {
            filter();
        }

        if (qs && qs.i) {
            setSelected(parseInt(qs.i, 10), true);
        } else {
            setSelected(0);
        }

        if (qs && (qs.i || qs.search)) {
            $search[0].focus();
        } else {
            document.body.focus();
        }
    }

    $search.keydown(function (e) {
        // Run First keydown
        setTimeout(function () { keypress(e); }, 0);

        // // Set up a timer to repeat (holding down arrow keys etc)
        var timer = null, timeout = null;
        var repeatKeys   = [KEY.PAGE_UP, KEY.PAGE_DOWN,
                            KEY.ARROW_UP, KEY.ARROW_DOWN,
                            KEY.BACKSPACE];
        var set_interval = function () {
            timer = setInterval(function () { keypress(e); }, 60);
        };

        if ($.inArray(e.keyCode, repeatKeys) === -1) {
            return;
        }

        function cleanup() {
            window.clearTimeout(timeout);
            $(document).unbind("keyup", cleanup);
            clearInterval(timer);
        }

        $search.bind("keyup", cleanup);
        timeout = window.setTimeout(set_interval, 300);
    });

    $search.focus(function () {
        $search[0].focus();
        $("#sidebar").removeClass("inactive");
        searchActive = true;
    });

    $search.blur(function () {
        $("#sidebar").addClass("inactive");
        searchActive = false;
        document.body.focus();
    });

    $(document).bind("keypress", function(e) {

        if (e.keyCode === KEY.TAB && !e.metaKey) {

            e.preventDefault();
            e.stopPropagation();

            if (searchActive) {
                $search.blur();
                $("#content")[0].focus();
            } else {
                $search[0].focus();
            }

            return false;
        }
        return true;
    });

    showingFuns = !!window.localStorage &&
        strToBool(window.localStorage.footer);

    if (document.title.match("Module Index") === null) {
        updateTitle();
        // setDetails(getDetails(), function() {
        //     $("#funwrapper").css({"display":"block"});
        // });
    }

    // $("#viewfuns").bind("mousedown", function(e) {

    //     showingFuns = !showingFuns;
    //     setDetails(getDetails(), null);

    //     if (!!window.localStorage) {
    //         window.localStorage.footer = showingFuns;
    //     }
    // });

    $(window).bind('resize', windowResize);
    windowResize();

    init();

})(index);

/* Copyright (C) 2007, 2008 gnombat@users.sourceforge.net */
/* License: http://shjs.sourceforge.net/doc/gplv3.html */

if(!this.sh_languages){this.sh_languages={}}var sh_requests={};function sh_isEmailAddress(a){if(/^mailto:/.test(a)){return false}return a.indexOf("@")!==-1}function sh_setHref(b,c,d){var a=d.substring(b[c-2].pos,b[c-1].pos);if(a.length>=2&&a.charAt(0)==="<"&&a.charAt(a.length-1)===">"){a=a.substr(1,a.length-2)}if(sh_isEmailAddress(a)){a="mailto:"+a}b[c-2].node.href=a}function sh_konquerorExec(b){var a=[""];a.index=b.length;a.input=b;return a}function sh_highlightString(B,o){if(/Konqueror/.test(navigator.userAgent)){if(!o.konquered){for(var F=0;F<o.length;F++){for(var H=0;H<o[F].length;H++){var G=o[F][H][0];if(G.source==="$"){G.exec=sh_konquerorExec}}}o.konquered=true}}var N=document.createElement("a");var q=document.createElement("span");var A=[];var j=0;var n=[];var C=0;var k=null;var x=function(i,a){var p=i.length;if(p===0){return}if(!a){var Q=n.length;if(Q!==0){var r=n[Q-1];if(!r[3]){a=r[1]}}}if(k!==a){if(k){A[j++]={pos:C};if(k==="sh_url"){sh_setHref(A,j,B)}}if(a){var P;if(a==="sh_url"){P=N.cloneNode(false)}else{P=q.cloneNode(false)}P.className=a;A[j++]={node:P,pos:C}}}C+=p;k=a};var t=/\r\n|\r|\n/g;t.lastIndex=0;var d=B.length;while(C<d){var v=C;var l;var w;var h=t.exec(B);if(h===null){l=d;w=d}else{l=h.index;w=t.lastIndex}var g=B.substring(v,l);var M=[];for(;;){var I=C-v;var D;var y=n.length;if(y===0){D=0}else{D=n[y-1][2]}var O=o[D];var z=O.length;var m=M[D];if(!m){m=M[D]=[]}var E=null;var u=-1;for(var K=0;K<z;K++){var f;if(K<m.length&&(m[K]===null||I<=m[K].index)){f=m[K]}else{var c=O[K][0];c.lastIndex=I;f=c.exec(g);m[K]=f}if(f!==null&&(E===null||f.index<E.index)){E=f;u=K;if(f.index===I){break}}}if(E===null){x(g.substring(I),null);break}else{if(E.index>I){x(g.substring(I,E.index),null)}var e=O[u];var J=e[1];var b;if(J instanceof Array){for(var L=0;L<J.length;L++){b=E[L+1];x(b,J[L])}}else{b=E[0];x(b,J)}switch(e[2]){case -1:break;case -2:n.pop();break;case -3:n.length=0;break;default:n.push(e);break}}}if(k){A[j++]={pos:C};if(k==="sh_url"){sh_setHref(A,j,B)}k=null}C=w}return A}function sh_getClasses(d){var a=[];var b=d.className;if(b&&b.length>0){var e=b.split(" ");for(var c=0;c<e.length;c++){if(e[c].length>0){a.push(e[c])}}}return a}function sh_addClass(c,a){var d=sh_getClasses(c);for(var b=0;b<d.length;b++){if(a.toLowerCase()===d[b].toLowerCase()){return}}d.push(a);c.className=d.join(" ")}function sh_extractTagsFromNodeList(c,a){var f=c.length;for(var d=0;d<f;d++){var e=c.item(d);switch(e.nodeType){case 1:if(e.nodeName.toLowerCase()==="br"){var b;if(/MSIE/.test(navigator.userAgent)){b="\r"}else{b="\n"}a.text.push(b);a.pos++}else{a.tags.push({node:e.cloneNode(false),pos:a.pos});sh_extractTagsFromNodeList(e.childNodes,a);a.tags.push({pos:a.pos})}break;case 3:case 4:a.text.push(e.data);a.pos+=e.length;break}}}function sh_extractTags(c,b){var a={};a.text=[];a.tags=b;a.pos=0;sh_extractTagsFromNodeList(c.childNodes,a);return a.text.join("")}function sh_mergeTags(d,f){var a=d.length;if(a===0){return f}var c=f.length;if(c===0){return d}var i=[];var e=0;var b=0;while(e<a&&b<c){var h=d[e];var g=f[b];if(h.pos<=g.pos){i.push(h);e++}else{i.push(g);if(f[b+1].pos<=h.pos){b++;i.push(f[b]);b++}else{i.push({pos:h.pos});f[b]={node:g.node.cloneNode(false),pos:h.pos}}}}while(e<a){i.push(d[e]);e++}while(b<c){i.push(f[b]);b++}return i}function sh_insertTags(k,h){var g=document;var l=document.createDocumentFragment();var e=0;var d=k.length;var b=0;var j=h.length;var c=l;while(b<j||e<d){var i;var a;if(e<d){i=k[e];a=i.pos}else{a=j}if(a<=b){if(i.node){var f=i.node;c.appendChild(f);c=f}else{c=c.parentNode}e++}else{c.appendChild(g.createTextNode(h.substring(b,a)));b=a}}return l}function sh_highlightElement(d,g){sh_addClass(d,"sh_sourceCode");var c=[];var e=sh_extractTags(d,c);var f=sh_highlightString(e,g);var b=sh_mergeTags(c,f);var a=sh_insertTags(b,e);while(d.hasChildNodes()){d.removeChild(d.firstChild)}d.appendChild(a)}function sh_getXMLHttpRequest(){if(window.ActiveXObject){return new ActiveXObject("Msxml2.XMLHTTP")}else{if(window.XMLHttpRequest){return new XMLHttpRequest()}}throw"No XMLHttpRequest implementation available"}function sh_load(language,element,prefix,suffix){if(language in sh_requests){sh_requests[language].push(element);return}sh_requests[language]=[element];var request=sh_getXMLHttpRequest();var url=prefix+"sh_"+language+suffix;request.open("GET",url,true);request.onreadystatechange=function(){if(request.readyState===4){try{if(!request.status||request.status===200){eval(request.responseText);var elements=sh_requests[language];for(var i=0;i<elements.length;i++){sh_highlightElement(elements[i],sh_languages[language])}}else{throw"HTTP error: status "+request.status}}finally{request=null}}};request.send(null)}function sh_highlightDocument(g,k){var b=document.getElementsByTagName("pre");for(var e=0;e<b.length;e++){var f=b.item(e);var a=sh_getClasses(f);for(var c=0;c<a.length;c++){var h=a[c].toLowerCase();if(h==="sh_sourcecode"){continue}if(h.substr(0,3)==="sh_"){var d=h.substring(3);if(d in sh_languages){sh_highlightElement(f,sh_languages[d])}else{if(typeof(g)==="string"&&typeof(k)==="string"){sh_load(d,f,g,k)}else{throw'Found <pre> element with class="'+h+'", but no such language exists'}}break}}}};

sh_languages['erlang'] = [
  [
    [
      /\b(?:div|default|rem|or|xor|bor|bxor|bsl|bsr|and|band|not|bnot|abs|alive|apply|atom_to_list|binary_to_list|binary_to_term|concat_binary|date|disconnect_node|element|erase|exit|float|float_to_list|get|get_keys|group_leader|halt|hd|integer_to_list|is_alive|length|link|list_to_atom|list_to_binary|list_to_float|list_to_integer|list_to_pid|list_to_tuple|load_module|make_ref|monitor_node|node|nodes|now|open_port|pid_to_list|process_flag|process_info|process|put|register|registered|round|self|setelement|size|spawn|spawn_link|split_binary|statistics|term_to_binary|throw|time|tl|trunc|tuple_to_list|unlink|unregister|whereis|atom|binary|constant|function|integer|list|number|pid|ports|port_close|port_info|reference|record|check_process_code|delete_module|get_cookie|hash|math|module_loaded|preloaded|processes|purge_module|set_cookie|set_node|acos|asin|atan|atan2|cos|cosh|exp|log|log10|pi|pow|power|sin|sinh|sqrt|tan|tanh|call|module_info|parse_transform|undefined_function|error_handler|after|begin|case|catch|cond|end|fun|if|let|of|query|receive|when|creation|current_function|dictionary|group_leader|heap_size|high|initial_call|linked|low|memory_in_use|message_queue|net_kernel|node|normal|priority|reductions|registered_name|runnable|running|stack_trace|status|timer|trap_exit|waiting|command|count_in|count_out|creation|in|in_format|linked|node|out|owner|packeting|atom_tables|communicating|creation|current_gc|current_reductions|current_runtime|current_wall_clock|distribution_port|entry_points|error_handler|friends|garbage_collection|magic_cookie|magic_cookies|module_table|monitored_nodes|name|next_ref|ports|preloaded|processes|reductions|ref_state|registry|runtime|wall_clock|apply_lambda|module_info|module_lambdas|record|record_index|record_info|badarg|nocookie|false|fun|true|badsig|kill|killed|exit|normal)\b/g,
      'sh_keyword',
      -1
    ],
    [
      /%/g,
      'sh_comment',
      1
    ],
    [
      /\(|\)|\{|\}|\[|\]|\||;|,|\?|#/g,
      'sh_normal',
      -1
    ],
    [
      /\b[+-]?(?:(?:0x[A-Fa-f0-9]+)|(?:(?:[\d]*\.)?[\d]+(?:[eE][+-]?[\d]+)?))u?(?:(?:int(?:8|16|32|64))|L)?\b|\$\x+/g,
      'sh_number',
      -1
    ],
    [
      /"/g,
      'sh_string',
      2
    ],
    [
      /\'/g,
      'sh_string',
      3
    ],
    [
      /\w+(?:\s+)?[:@](?:\s+)?\w+/g,
      'sh_function',
      -1
    ],
    [
      /-compile|-define|-else|-endif|-export|-file|-ifdef|-ifndef|-import|-include|-include_lib|-module|-record|-undef|-author|-copyright|-doc/g,
      'sh_preproc',
      -1
    ],
    [
      /\+|-|\*|\/|==|=|=:=|=\/=|<|=<|>|>=|\+\+|--|=|!|<-|->|:|_|@|\\|\"|\./g,
      'sh_symbol',
      -1
    ]
  ],
  [
    [
      /$/g,
      null,
      -2
    ]
  ],
  [
    [
      /\\(?:\\|")/g,
      null,
      -1
    ],
    [
      /"/g,
      'sh_string',
      -2
    ]
  ],
  [
    [
      /\\(?:\\|\')/g,
      null,
      -1
    ],
    [
      /\'/g,
      'sh_string',
      -2
    ]
  ]
];
sh_highlightDocument();
